### ISSUE TITLE
Large for-loops in property expression causes StackOverflowException

### ISSUE THREAD

**Issue by marklam** [hedgehogqa/fsharp-hedgehog#177]
This is probably due to something not being tail-recursive.
A large for loop doesn't overflow the `seq` expression, but it does overflow `property`

```f#
open Hedgehog

// Learn more about F# at http://fsharp.org
// See the 'F# Tutorial' project for more help.

[<EntryPoint>]
let main argv = 
    let values1 = seq {
            for a = 0 to System.Int32.MaxValue do
                ()
            
            yield true
        }

    let values2 = property {
            for a = 0 to System.Int32.MaxValue do
                ()
            
            return true
        }

    printfn "Checking Sequence"
    printfn "Sequence %A" (values1 |> Seq.last)
    printfn "Checking Hedgehog"
    printfn "Hedgehog %A" (values2 |> Property.check)
    0 // return an integer exit code
```

The call stack contains *many* repeated sequences of
```
 	Hedgehog.dll!Hedgehog.Gen.bindRandom@40<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>, System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>>.Invoke(Hedgehog.Seed seed0, int size = 1) Line 41	F#
 	Hedgehog.dll!Hedgehog.Random.map@39-2<Hedgehog.Tree<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>>, Hedgehog.Tree<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>>>.Invoke(Hedgehog.Seed seed, int size) Line 40	F#
 	Hedgehog.dll!Hedgehog.Tree.bind<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>, System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>>(Hedgehog.Tree<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>> _arg1, Microsoft.FSharp.Core.FSharpFunc<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>, Hedgehog.Tree<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>>> k = {Hedgehog.Gen.bindRandom@47-1<System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>, System.Tuple<Hedgehog.Journal, Hedgehog.Result<Microsoft.FSharp.Core.Unit>>>}) Line 31	F#
```

Since the loop I had in my real test case was over a set of items, and contained no `let!` or similar, I was able to use `Set.iter` as a workaround, but I thought it was worth reporting.

**Comment by moodmosaic** [comment]
If I have to guess, it's the stuff in `LazyList` and/or `Gen.bind` that's not tail recursive. I'd be curious to try using `seq` instead of `LazyList` making sure it's evaluated only once.

Thank you for reporting this 👍

**Comment by Porges** [comment]
I did convert from `LazyList` to `seq` as an experiment and it wasn't that hard 🙂

**Comment by moodmosaic** [comment]
Feel free to send a PR if it's worth it. :wink:

**Comment by Porges** [comment]
I have a fix for blowing the stack, but there's still an issue in that the current implementation makes this very slow. Investigating more to see if we can improve it...

**Comment by moodmosaic** [comment]
Great! Today I'm traveling but if you set something up, I'll be able to look into it tomorrow.

**Comment by Porges** [comment]
Actually I haven't fully fixed it; can still blow the stack it just takes a lot longer ... will need more work.

**Comment by Porges** [comment]
So just to share what I was trying: one issue is that consecutive `bind`s generate a large left-leaning composition of `bind`s which means that there's a big stack of `Random`s holding onto other `Random`s (inside `bindRandom`). Attempting to execute this big tower of `Random`s then overflows the stack. (You can replicate this more easily by folding `bind` over a sequence of `Property`s.)

One fix for this is to reassociate the binds to always be right-leaning, so that the implementation can be tail recursive and will never blow the stack. This is what I'm attempting at the moment: I make no guarantees 😀

**Comment by moodmosaic** [comment]
>This is what I'm attempting at the moment: I make no guarantees

^^ 😆 🚀 

Sure it'd be absolutely useful to have this optimized/rewritten! In the long run, though, I think the 'right' way would be to get rid of `Random.fs` entirely...

---

Getting rid of `Random.fs` means it should all be in `Gen.fs` similar to the [Haskell version](https://github.com/hedgehogqa/haskell-hedgehog/blob/master/hedgehog/src/Hedgehog/Internal/Gen.hs) of `Gen` and that would allow us to add [Gen.freeze](http://hackage.haskell.org/package/hedgehog/docs/Hedgehog-Gen.html#v:freeze) and friends, so basically `Gen` would be defined as

```f#
type Gen<'a> =
    | Gen of (Size -> Seed -> Tree<Option<'a>>)
```

(`Range` `Shrink` and `Seed` are already in-line with the Haskell version but `Tree` isn't.)

**Comment by TysonMN** [comment]
What is the meaning of the `option` states in this alternative definition of `Gen`?

**Comment by moodmosaic** [comment]
`Some` would be the list of children and `None` would mean cut the tree off from this point, so that you can't keep trying to shrink.

**Comment by moodmosaic** [comment]
And probably what we'd want is actually flipped, `Option<Tree<'a>>`, though its been a while since that discussion.

**Comment by TysonMN** [comment]
I still don't understand either of `Option<Tree<'a>>` and `Tree<Option<'a>>`.

Maybe the type `Option<Tree<'a>>` could represent the concept of filtering.  By the time we are sinking though, a value failed the test, so we must be in the `Some` state.  Thus, shrinking is unchanged for this type.

The type `Tree<Option<'a>>` means each value in the tree might or might not be there.  In either case, every node still has a list of children.

Neither of these types corresponds to having or not having a list of children.

Shirking currently stops when none of children values cause the test to fail.  This is vaculously true when the list is empty.

**Comment by adam-becker** [comment]
@TysonMN Is this related to the work you've done in #289?

**Comment by TysonMN** [comment]
Probably not.  However, I expect this is caused by `Random.bind` having a similar same problem as `Random.map`.  I don't know yet, but I hope the difference list technique will also help fix this problem.

**Comment by adam-becker** [comment]
I don't know if this is still an issue. I tried running the benchmarks with `10,000` cases and it didn't overflow, the comment on the benchmark said it would with that value. Closing for now, if this re-emerges please open a new issue.

**Comment by TysonMN** [comment]
I also tried some overflow tests lately, and I was a bit confused that I was not getting overflows.  Even if 10,000 would cause an overflow, it is not that important.  I don't expect anyone to get anywhere near the overflow limit in practice.  As such, this work was mostly of theoretical interest.  I am fine having this closed.
